### PR TITLE
Update child modules docs

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Part of docs revamp #7666

Moved child modules docs to LightningModule docs.
Also improved the sidebar for LightningModule doc.

Before
Screenshot 2021-12-21 at 5 25 18 PM

After
Screenshot 2021-12-21 at 5 25 42 PM

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7 @carmocca @justusschock @awaelchli @ananthsub @ninginthecloud @jjenniferdai

### CODE REVIEW

File: docs/source/common/child_modules.rst [hunk 1]
```diff
@@ -38,7 +14,7 @@ The only things that change in the `Autoencoder` model are the init, forward, tr
         pass
 
 
-    class AutoEncoder(LitMNIST):
+    class AutoEncoder(LightningModule):
```
Comment (tchaton) [inline docs/source/common/child_modules.rst:17]: Let's re-organize this example as a System.

Comment (tchaton) [reply docs/source/common/child_modules.rst:17]:
```py
class AutoEncoder(nn.Module):

    def __init__(self):
        self.encoder = Encoder()
        self.decoder = Decoder()

    def forward(self, x):
        return self.decoder(self.encoder(x))


class Model(pl.LightningModule):

    def __init__(self, auto_encoder):
        # A single nn.Module is being set to the LightningModule
        self.auto_encoder = auto_encoder
        self.metric = MSE()

    def forward(self, x):
        return self.auto_encoder.encoder(x)

    def training_step(self, batch, batch_idx):
        x, _ = batch
        x_hat = self.auto_encoder(x)
        loss = self.metric(x, x_hat)
        return loss

lightning_module = Model(AutoEncoder())
```

Comment (tchaton) [reply docs/source/common/child_modules.rst:17]: Would be good to adapt this methodology over the codebase.

File: docs/source/common/child_modules.rst [hunk 2]
```diff
@@ -72,13 +48,13 @@ The only things that change in the `Autoencoder` model are the init, forward, tr
             self.log(f"{prefix}_loss", loss)
 
 
-and we can train this using the same trainer
+and we can train this using the same Trainer instance:
```
Comment (tchaton) [inline docs/source/common/child_modules.rst:51]: A bit confusing to me.

File: docs/source/common/lightning_module.rst [hunk 3]
```diff
@@ -191,8 +195,9 @@ Under the hood, Lightning does the following (pseudocode):
         optimizer.step()
 
 
-Training Epoch-Level Metrics
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Train Epoch-Level Metrics
+=========================
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:199]:
```suggestion
Epoch-Level Metrics during Training
==========================
```

This title is a bit confusing

Comment (tchaton) [reply docs/source/common/lightning_module.rst:199]: Might be better to have Epoch-Level Metrics and sub section for training, validation and test.

File: docs/source/common/lightning_module.rst [hunk 4]
```diff
@@ -484,116 +503,56 @@ There are two ways to call ``test()``:
...
+Inference
+*********
 
-        def configure_optimizers(self):
-            return torch.optim.Adam(self.parameters(), lr=0.0002)
+Prediction Loop
+===============
 
-Which can be trained like this:
+To activate the prediction loop, you may need to override the :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:513]:
```suggestion
By default, the :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method runs under the hook the  :meth:`~pytorch_lightning.core.lightning.LightningModule.forward`. In order to customize this behaviour, simply overrides the  :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method.
```

It might be worth to add an example with `MCDropout` where predict_step is overridden for example

File: docs/source/common/lightning_module.rst [hunk 5]
```diff
@@ -484,116 +503,56 @@ There are two ways to call ``test()``:
...
-The methods above are part of the LightningModule interface:
-
-- training_step
-- validation_step
-- test_step
-- predict_step
-- configure_optimizers
+    class LitModel(pl.LightningModule):
+        def predict_step(self, batch, batch_idx):
+            x, y = batch
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:520]: Let's show a MCDropout example here and a link to the method usage.

File: docs/source/common/lightning_module.rst [hunk 6]
```diff
@@ -731,13 +691,90 @@ a ``LightningModule``.
     with torch.no_grad():
         y_hat = model(x)
 
+
+-----------
+
+
+*************
+Child Modules
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:699]: I am not sure we want to advise this anymore as it is better to follow the design pattern to simplify using nn.Module without Lightning.

File: docs/source/common/lightning_module.rst [hunk 7]
```diff
@@ -731,13 +691,90 @@ a ``LightningModule``.
...
+
+and we can train this using the same Trainer instance:
+
+.. code-block:: python
+
+    autoencoder = AutoEncoder()
+    trainer = Trainer()
+    trainer.fit(autoencoder, train_dataloader, val_dataloader)
+
+And remember that the forward method should define the practical use of a LightningModule.
```
Comment (Borda) [inline docs/source/common/lightning_module.rst:760]:
```suggestion
And remember that the forward method should define the practical use of a ``LightningModule``.
```

File: docs/source/common/lightning_module.rst [hunk 8]
```diff
@@ -731,13 +691,90 @@ a ``LightningModule``.
...
+        def _shared_eval(self, batch, batch_idx, prefix):
+            x, _ = batch
+            representation = self.encoder(x)
+            x_hat = self.decoder(representation)
+
+            loss = self.metric(x, x_hat)
+            self.log(f"{prefix}_loss", loss)
+
+
+and we can train this using the same Trainer instance:
```
Comment (Borda) [inline docs/source/common/lightning_module.rst:752]:
```suggestion
and we can train this using the same ``Trainer`` instance:
```

File: docs/source/common/lightning_module.rst [hunk 9]
```diff
@@ -731,13 +691,90 @@ a ``LightningModule``.
...
+
+.. code-block:: python
+
+    autoencoder = AutoEncoder()
+    trainer = Trainer()
+    trainer.fit(autoencoder, train_dataloader, val_dataloader)
+
+And remember that the forward method should define the practical use of a LightningModule.
+In this case, we want to use the `AutoEncoder` to extract image representations
```
Comment (Borda) [inline docs/source/common/lightning_module.rst:761]:
```suggestion
In this case, we want to use the ``AutoEncoder`` to extract image representations
```

File: docs/source/common/lightning_module.rst (hunk 3 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:199]: Right now ste structure is like:
```
- Training
    - ...
    - epoch level metrics
    - ...
- Validation
    - ...
    - epoch level metrics
    - ...
- Testing
    - ... 
```
which I think is better because we have other subsections as well that belongs to each of the main heading.

File: docs/source/common/child_modules.rst (hunk 2 above)
Comment (rohitgr7) [reply docs/source/common/child_modules.rst:51]: Ah! it's incorrect. Removed.

File: docs/source/common/lightning_module.rst (hunk 6 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:699]: I think it's okay to have as long as all are under `nn.Module`. It represents that LightningModule is just ``nn.Module`` and one can bring modules from outside.

File: docs/source/common/lightning_module.rst (hunk 5 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst]: done!

**Comment by kaushikb11** [comment]
Loved the improved side bar! Much cleaner.

File: docs/source/common/child_modules.rst [hunk 10]
```diff
@@ -65,23 +47,22 @@ The only things that change in the `Autoencoder` model are the init, forward, tr
...
 
 
-and we can train this using the same trainer
+and we can train this using the ``Trainer``:
 
 .. code-block:: python
 
-    autoencoder = AutoEncoder()
+    model = AutoEncoder()
+    autoencoder = LitAutoEncoder(model)
```
Comment (kaushikb11) [inline docs/source/common/child_modules.rst:60]: Should we switch the variable names?

File: docs/source/common/lightning_module.rst (hunk 3 above)
Comment (kaushikb11) [reply docs/source/common/lightning_module.rst:199]: @rohitgr7 Could you update the section titles?

**Review by kaushikb11 — APPROVED** [review]

File: docs/source/common/lightning_module.rst [hunk 11]
```diff
@@ -482,118 +509,76 @@ There are two ways to call ``test()``:
...
-
-            # encode
-            x = x.view(x.size(0), -1)
-            z = self.encoder(x)
+.. note::
 
-            # decode
-            recons = self.decoder(z)
+    It is recommended to test on single device since Distributed Training such as DDP internally
+    uses :class:`~torch.utils.data.distributed.DistributedSampler` which replicates some samples to
```
Comment (kaushikb11) [inline docs/source/common/lightning_module.rst:515]: > replicated some samples?

It could worded better.

File: docs/source/common/child_modules.rst (hunk 10 above)
Comment (tchaton) [inline docs/source/common/child_modules.rst:60]:
```suggestion
    autoencoder = AutoEncoder()
    lightning_module = LitAutoEncoder(autoencoder)
```

File: docs/source/common/lightning_module.rst [hunk 12]
```diff
@@ -373,8 +385,17 @@ and calling :meth:`~pytorch_lightning.trainer.trainer.Trainer.validate`.
     trainer = Trainer()
     trainer.validate(model)
 
+.. note::
+
+    It is recommended to validate on single device since Distributed Training such as DDP internally
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:390]: Neat ! Let's make this a warning.

File: docs/source/common/lightning_module.rst [hunk 13]
```diff
@@ -482,118 +509,76 @@ There are two ways to call ``test()``:
...
+===============
 
-        def configure_optimizers(self):
-            return torch.optim.Adam(self.parameters(), lr=0.0002)
+By default, the :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method runs the
+:meth:`~pytorch_lightning.core.lightning.LightningModule.forward` method. In order to customize this behaviour,
+simply override the :meth:`~pytorch_lightning.core.lightning.LightningModule.predict_step` method.
 
-Which can be trained like this:
+For the example let's override ``predict_step`` and try out `Monte Carlo Dropout <https://arxiv.org/pdf/1506.02142.pdf>`_:
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:532]: Neat !

File: docs/source/common/lightning_module.rst [hunk 14]
```diff
@@ -482,118 +509,76 @@ There are two ways to call ``test()``:
...
 
-    autoencoder = Autoencoder()
-    trainer = pl.Trainer(gpus=1)
-    trainer.fit(autoencoder, train_dataloader, val_dataloader)
-
-This simple model generates examples that look like this (the encoders and decoders are too weak)
-
-.. figure:: https://pl-bolts-doc-images.s3.us-east-2.amazonaws.com/pl_docs/ae_docs.png
-    :width: 300
+    class LitModel(pl.LightningModule):
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:536]:
```suggestion
    class LitMCdropoutModel(pl.LightningModule):
```

File: docs/source/common/lightning_module.rst [hunk 15]
```diff
@@ -482,118 +509,76 @@ There are two ways to call ``test()``:
...
+        def __init__(self):
+            super().__init__()
+            self.linear = nn.Linear(4, 5)
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:539]:
```suggestion
        def __init__(self, model, mc_iteration: int):
            super().__init__()
            self.linear = model
            self.mc_iteration = mc_iteration
```

File: docs/source/common/lightning_module.rst [hunk 16]
```diff
@@ -482,118 +509,76 @@ There are two ways to call ``test()``:
...
+            # enable Monte Carlo Dropout
+            self.dropout.train()
 
-- training_step
-- validation_step
-- test_step
-- predict_step
-- configure_optimizers
+            # take average of 7 iterations
+            pred = torch.vstack([self.dropout(self.linear(x)).unsqueeze(0) for _ in range(7)]).mean(dim=0)
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:547]:
```suggestion
            pred = torch.vstack([self.dropout(self.linear(x)).unsqueeze(0) for _ in range(self.mc_iteration)]).mean(dim=0)
```

Comment (tchaton) [reply docs/source/common/lightning_module.rst:547]: Usually, it is more some layers, dropout, layer, ...

File: docs/source/common/lightning_module.rst [hunk 17]
```diff
@@ -482,118 +509,76 @@ There are two ways to call ``test()``:
...
+    class LitModel(pl.LightningModule):
+        def __init__(self):
+            super().__init__()
+            self.linear = nn.Linear(4, 5)
+            self.dropout = nn.Dropout()
 
-The methods above are part of the LightningModule interface:
+        def predict_step(self, batch, batch_idx):
+            # enable Monte Carlo Dropout
+            self.dropout.train()
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:544]:
```suggestion
            self.train()
```

File: docs/source/common/lightning_module.rst [hunk 18]
```diff
@@ -482,118 +509,76 @@ There are two ways to call ``test()``:
...
-            self.log("val_loss", loss)
+    # disable grads + batchnorm + dropout
+    torch.set_grad_enabled(False)
+    model.eval()
+    all_preds = []
 
-        def shared_step(self, batch):
-            x, _ = batch
+    for batch in predict_dataloader:
+        pred = model.predict_step()
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:560]:
```suggestion
    for batch_idx, batch in enumerate(predict_dataloader):
        pred = model.predict_step(batch, batch_idx)
```

File: docs/source/common/lightning_module.rst [hunk 19]
```diff
@@ -1109,10 +1094,9 @@ Set and access example_input_array, which basically represents a single batch.
         # generate some images using the example_input_array
         gen_images = self.generator(self.example_input_array)
 
---------------
-
 datamodule
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:1097]: This is deprecated right ?

**Review by tchaton — APPROVED** [review]
Looks great ! I would recommend writing down `System`.

File: docs/source/common/lightning_module.rst [hunk 20]
```diff
@@ -918,21 +915,19 @@ The current epoch
         if self.current_epoch == 0:
             ...
 
--------------
```
Comment (awaelchli) [inline docs/source/common/lightning_module.rst:921]: Are we removing these because we are  updating the styling and spacing globally in the docs?

So far @williamFalcon  strongly insisted in this particular formatting style, so we should keep it unless clearly agreed on.

File: docs/source/common/child_modules.rst (hunk 10 above)
Comment (rohitgr7) [reply docs/source/common/child_modules.rst:60]: yes. will update

File: docs/source/common/lightning_module.rst (hunk 17 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:544]: dropout is the only one that needs to be kept in training mode.

File: docs/source/common/lightning_module.rst (hunk 19 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:1097]: yes..thanks :)

File: docs/source/common/lightning_module.rst (hunk 20 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:921]: yes.. mostly the top-level headings are separated by these breaks. Adding it to all of these individual properties adds up a lot of space/padding in between. Without them, it looks good to me.
Screenshot 2022-01-10 at 1 46 23 AM

File: docs/source/common/lightning_module.rst [hunk 21]
```diff
@@ -253,9 +259,9 @@ The matching pseudocode is:
 .. code-block:: python
 
     outs = []
-    for batch in train_dataloader:
+    for batch_idx, batch in enumerate(train_dataloader):
         # forward
-        out = training_step(val_batch)
+        out = training_step(batch, batch_idx)
```
Comment (akihironitta) [inline docs/source/common/lightning_module.rst:264]: [same as above] `loss.backward()` below needs updating to `out.backward()`.

File: docs/source/common/lightning_module.rst [hunk 22]
```diff
@@ -482,118 +510,79 @@ There are two ways to call ``test()``:
...
+        def predict_step(self, batch, batch_idx):
+            # enable Monte Carlo Dropout
+            self.dropout.train()
 
-- training_step
-- validation_step
-- test_step
-- predict_step
-- configure_optimizers
+            # take average of 7 iterations
```
Comment (akihironitta) [inline docs/source/common/lightning_module.rst:550]:
```suggestion
            # take average of `self.mc_iteration` iterations
```

File: docs/source/common/child_modules.rst [hunk 23]
```diff
@@ -38,22 +14,28 @@ The only things that change in the `Autoencoder` model are the init, forward, tr
...
+
+        def forward(self, x):
+            return self.decoder(self.encoder(x))
+
+
+    class LitAutoEncoder(LightningModule):
+        def __init__(self, auto_encoder):
+            super().__init__()
+            self.auto_encoder = auto_encoder
             self.metric = MSE()
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/common/child_modules.rst]: let's use one from torch directly.

File: docs/source/common/lightning_module.rst [hunk 24]
```diff
@@ -213,9 +218,9 @@ requested metrics across a complete epoch and devices. Here's the pseudocode of
 .. code-block:: python
 
     outs = []
-    for batch in train_dataloader:
+    for batch_idx, batch in enumerate(train_dataloader):
         # forward
-        out = training_step(val_batch)
+        out = training_step(batch, batch_idx)
         outs.append(out)
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst]: I swear, I read this example twice 😄

**Review by akihironitta — APPROVED** [review]
